### PR TITLE
propose vim as manpager that syntax highlights and follows symlinks

### PR DESCRIPTION
**Pull request by Konfekt**

### CODE REVIEW

File: runtime/plugin/manpager.vim [hunk 1]
```diff
@@ -0,0 +1,18 @@
+if empty($MAN_PN)
```
Comment (ZyX-I) [inline runtime/plugin/manpager.vim:5]: Missing header (see other runtime files).

File: runtime/plugin/manpager.vim [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+command! -nargs=0 MANPAGER call s:MANPAGER() | delcommand MANPAGER
+
+fun! s:MANPAGER()
+  bwipe!
+
+  let page = expand('$MAN_PN')
+  let page_pattern = '\v\w+(\-\w+)*'
+  let sec_pattern = '\v\(' . '\w+(\+\w+)*' . '\)'
+  let sec = substitute(matchstr(page, sec_pattern), '\v^\(|\)$', '', 'g')
+  let page = substitute(page, sec_pattern, '', '')
```
Comment (ZyX-I) [inline runtime/plugin/manpager.vim]: I would highly suggest  to use `matchlist`.

File: runtime/plugin/manpager.vim [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+
+  setlocal filetype=man
+  exe 'Man ' . sec . ' ' . page
```
Comment (ZyX-I) [inline runtime/plugin/manpager.vim]: `execute 'Man' sec page` is enough. Inserting spaces and concatenation looks better when implicit.

**Comment by chrisbra** [comment]
what is the difference with e.g. http://www.drchip.org/astronaut/vim/index.html#MANPAGEVIEW?

**Comment by ZyX-I** [comment]
@chrisbra I think it is obvious: whan you linked there is a _plugin_. If this is merged it would be part of the standard distribution. Given that all necessary functionality is already here it is more convenient to have this built in.

**Comment by Konfekt** [comment]
@chrisbra There are a couple of manpager plugins but the most solid one is already built-in, short of these few lines of Vimscript. People should know about it.

File: runtime/plugin/manpager.vim (hunk 2 above)
Comment (Konfekt) [inline runtime/plugin/manpager.vim]: matchlist instead of matchstr ? `:Man` only handles a single man page as input by `:help Man`.

Comment (ZyX-I) [inline runtime/plugin/manpager.vim]: @Konfekt How is this related? `matchlist` allows you to perform pattern matching only once, and also show how `$MAN_PN` looks like.

**Comment by brammool** [comment]
Where does this work?  Only in bash perhaps?

**Comment by Konfekt** [comment]
Works fine in bash and zsh.

**Comment by Konfekt** [comment]
If you refer to the doc, it's just that I only knew the configuration files for bash and zsh. But  if kornshell and friends accept a manpager variable, then for these it works probably as well?!

**Comment by Konfekt** [comment]
If someone could check if this works on kornshell, dash, tcsh on friends? Probably fish, along with `zsh` and `bash`, tested over here,  is fine.

**Comment by muru** [comment]
`export foo=bar` works on ksh and dash. For (t)csh, you'll need to use `setenv` (see this nice post on [SuperUser.com](http://superuser.com/a/373964/334516)):

```
setenv MANPAGER "vim -M +MANPAGER -"
```

[For fish](http://fishshell.com/docs/current/tutorial.html#tut_exports), apparently you have to use `set -x`:

```
set -x MANPAGER "vim -M +MANPAGER -"
```

Of course, it depends on whether the `man` command accepts the `MANPAGER` variable. [POSIX](http://pubs.opengroup.org/onlinepubs/9699919799/utilities/man.html) only specifies `PAGER`, but [man-db](http://www.nongnu.org/man-db/) (the implementation seen on most Linux systems) and [FreeBSD's `man`](https://www.freebsd.org/cgi/man.cgi?query=man) both support `MANPAGER`, so I guess it's fairly widely supported.

The more important question is whether other `man` implementations set `MAN_PN`. And it seems [at least FreeBSD doesn't](http://chat.stackexchange.com/rooms/26/conversation/discovering-man), and it provides no way to know what manpage was opened. 😭 Looks like this is Linux-specific at the moment (or wherever man-db is used).

Since `MAN_PN` is only needed to get the first manpage, you could perhaps use the old trick of removing backspaces, and then setting filetype to `man`. (And this again is not guaranteed to work if `man` uses something else to do the highlighting, like ANSI escape codes.)

**Comment by Konfekt** [comment]
Thanks a bunch ! Exactly, the definition of MAN_PN is the weak point. So after updating instructions for other shells, apparemment BSD is unsupported. Could you explain tha ol trick about the removal of backspaces that yet works around it?

**Comment by vim-ml** [comment]
- Konfekt vim-dev-github@256bit.org [2016-04-28 02:25 -0700]:

> Thanks a bunch ! Exactly, the definition of MAN_PN is the weak point. So after updating instructions for other shells, apparemment BSD is unsupported. Could you explain tha ol trick about the removal of backspaces that yet works around it? 

Do you mean:

stty erase <BS>

## Elimar

  Obviously the human brain works like a computer.
  Since there are no stupid computers humans can't be stupid.
  There are just a few running with Windows or even CE ;-)

**Comment by muru** [comment]
Typically man pages have bold and underlines displayed using `x\bx` and `_\bx`, `\b` being the backspace and `x` the letter in question (Thomas Dickey talks about this in [this Unix & Linux post](http://unix.stackexchange.com/a/271553/70524)). That's what makes using just `vim -` as the manpager annoying. The workaround I have seen posted in a number of places is to filter out the backspaces, and the characters that would have been printed over. Typically:

```
MANPAGER='col -b | vim -'
```

There's no reason to do this with `col -b`, of course. You could do something like:

``` viml
if empty($MAN_PN)
    autocmd StdinReadPost * set ft=man | call PrepManPager()
endif
```

And `PrepManPager()` defined in some suitable place:

``` viml
function! PrepManPager()
    setlocal modifiable
    %s/.\b\(.\)/\1/g
    1
    setlocal nomodified
    setlocal nomodifiable
endfunction
```

So, for the first manpage, you don't have to bother with manually finding the manpage being opened, since `man` itself provides that page to Vim. It just needs to be cleaned up. It is possible that `man` could be using ANSI escape codes, if `GROFF_SGR=1` is set in the environment and you're using man-db's `man`. In that case, you could do `:%s/\e\[\d\+m//g` instead.

---

Aside:

In the Unix & Linux post mentioned above, I'd posted my method, which involves keeping the backspaces and concealing them, instead of removing them. Personally, I feel the loss of functionality in searching is balanced by the added syntax highlighting I can do with the backspaces retained. There are manpages for which the standard `syntax/man.vim` is useless, because it has no way to know what the actual page would have highlighted. (Pick any C++ STL manpage, for example.) If anyone's interested, [here's how I use conceal for this](https://github.com/murukeshm/vim-manpager/blob/master/after/syntax/man.vim).

**Comment by Konfekt** [comment]
Ok, the documentation is updated according to the setup @murukeshm gave. Thanks again! Somehow

https://www.freebsd.org/cgi/man.cgi?query=man&manpath=SuSE+Linux/i386+11.3#end

suggests that MAN_PN might be defined on FreeBSD after all, therefore it's included nevertheless for the moment.

I did not really understand the backspace issue yet, as I thought it to be the merit of :Man (and thereby SungHyun Nam) to take care of these? The manpager.vim supplied only invokes it according to $MAN_PN, and already there stays out of the game.

**Comment by muru** [comment]
Oh, have a closer look at the manpage URL.  :) FreeBSD's manpage site hosts pages from a number of systems, including a few Linux distros, other BSDs and Unix-likes. The page you linked to is for SuSE, which is (was) a Linux distro.

The backspace suggestion was just to show that you could just use the page that `man` provides, even if `MAN_PN` isn't set.

**Comment by Konfekt** [comment]
Thanks, that was sloppy to overlook from my part. Ok, now you think that Vim can detect that it was started as MANPAGER and use the arguments passed to invoke the suitable `:Man` command? It seemed to me that both of these were only possible by `$MAN_PN`, 
- knowing that vim is called as MANPAGER, and
- the man page and its section.

A quick check of`v:progname/path` yields `vim`.

**Comment by muru** [comment]
No, I'm saying that we don't need to call `:Man` at all when starting Vim. `man` has already provided the manpage to Vim via stdin, so we don't need to know the name or section. We just need to clean up the input from `man`, and set the filetype to `man`.

Say, on a FreeBSD system, we have `MANPAGER='env MAN_PN=1 vim -'`. Now, `MAN_PN` will only be set if `vim` was invoked by `man`. Then we can check for `MAN_PN` and tidy up the input from `man`.

**Comment by Konfekt** [comment]
Thanks for the `env $MAN_PN=1` hint. The code and instructions are now updated to work on many POSIX compatibe systems (in particular FreeBSD). Untested are still AmigaOS, BeOS and OS/2. Should be ready for merge.

**Comment by verbitan** [comment]
Hi all,

This isn't working for me, I get the following error and a blank file.

```
Cannot find a 'VIM'.
Press ENTER or type command to continue
```

Pressing :ls shows the following.

```
:ls
  2 %a-  "[No Name]"                    line 1
```

And quitting Vim shows this on the command line.

```
man vim
Vim: Reading from stdin...
```

I have Vim 7.4.1816 installed, am running tcsh with MANPAGER set as `MANPAGER=env MAN_PN=1 vim -M +MANPAGER -`.

Am I missing something? Hopefully I just haven't configured something correctly.

**Comment by Konfekt** [comment]
Does first of all`:set ftplugin man` and `:Man vim` work fine? 

Might be that the issue is with the `:Man` command rather than the simplistic `:MANPAGER`. For example, `:Man` assumes that the `man` executable is  `\usr\bin\man`.  I proposed a more flexible approach to SungHyun Nam. After all, why not leave it `man`?

**Comment by verbitan** [comment]
If I invoke vim without a file `vim` then type `:Man vim` I get `E492: Not an editor command: Man vim`. I _think_ this is expected.

If I invoke vim without a file `vim` then type `:set filetype=man` followed by `:Man vim` it opens correctly. On my system (RHEL 6.5) `man` is at `/usr/bin/man`.

It seems the issue is only when invoking via `tcsh`. I also gave it a quick test in `bash` with the same result.

**Comment by ches** [comment]
I have problems similar to @verbitan's, `manpager.vim` actually seems to be breaking `man.vim` for me.

If `MAN_PN=1` is set in my shell, so that `manpager.vim` loads, starting Vim plainly and then running `:Man wc` explodes, but if I `unset MAN_PN` in the shell before starting Vim, `:Man` works fine.

Here's what I get in a window after running `:Man wc` if `MAN_PN` is set:

```
Vim: Warning: Output is not to a terminal
Vim: Warning: Input is not from a terminal

?1000h?1049h?1h27mm38;5;188m48;5;234mH2J?25l24;1H"-stdin-" 80L, 3587C?1000lVim: Reading from stdin...
?1000h1;24r?12;25h?12l?25h27mm38;5;188m48;5;234mH2J?1000lVim: Reading from stdin...
?1000h24;1H<n/jn2hxmmx7csbzj1ft7j7f3v40000gn/T/hg-editor-J64zyu.txt" [New File]?1000lVim: Reading from stdin...
?1000h
97m41mError detected while processing function <SNR>171_MANPAGER:m38;5;188m48;5;234m
93mline   16:m38;5;188m48;5;234m
97m41mE684: list index out of range: 3m38;5;188m48;5;234m
97m41mE15: Invalid expression: page_sec[3] page_sec[1]?1000lVim: Reading from stdin...
```

I'm on OS X using bash. I've started Vim with `vim -N -u NONE` and done `:runtime ftplugin/man.vim`, followed by `:Man wc`.

**Comment by Konfekt** [comment]
Ok, so does removing the line env MAN_PN=1 restore functionality?

**Comment by ches** [comment]
> Ok, so does removing the line env MAN_PN=1 restore functionality?

It fixes `:Man`, but that means `manpager.vim` is unusable since it uses presence of the env variable as a load guard.

**Comment by Konfekt** [comment]
But only on FreeBSD. Thing is, almost all Linux distributions use man-db which sets MAN_PN itself. The env command was meant as workaround for those that do not use it,  apparently FreeBSD, but not necessary for all the other. See the last line of :help manpager. Apparently this workaround is flawed.

**Comment by Konfekt** [comment]
So the instructions should be updated to propose the Version without env...  First, and then suggest env... as workaround if that fails. Maybe the two variants are exclusive. If you could give feedback? In particular someone on FreeBSD if the env..  variant works.

**Comment by verbitan** [comment]
I think I tried it without setting the environment variable too without success, but I'll double check tomorrow and post the results.

**Comment by Konfekt** [comment]
Ok, apparently for@ches it works. Confusing.

**Comment by verbitan** [comment]
Ok so, my environment is now configured as `MANPAGER=vim -M +MANPAGER -` which gives me the following error.

```
Error detected while processing command line:
E492: Not an editor command: MANPAGER
Press ENTER or type command to continue
```

Pressing enter actually displays the man page which is an improvement over before, however the filetype isn't set correctly as man.

```
set ft
filetype=
```

Typing `:Man vim` shows `E492: Not an editor command: Man vim`, typing `:set ft=man` followed by `:Man vim` works as per my previous comment.

Finally I tried setting `MANPAGER=vim -u NONE -M +MANPAGER -` just in case there was something in my vim configuration that was causing the issue, but I got the same result.

**Comment by verbitan** [comment]
A little more playing seems so indicate that this is only a problem for certain manpages. For example, the following don't work.

```
man echo
Cannot find a 'ECHO'.
Press ENTER or type command to continue
```

```
man vim
Cannot find a 'VIM'.
Press ENTER or type command to continue
```

Whereas these ones are working fine.

```
man man
man yum
```

The only thing I can see different between the two is that the ones that break start with uppercase command names in the manpage...

```
ECHO(1)                          User Commands                         ECHO(1)

NAME
       echo - display a line of text
```

```
VIM(1)                                                                  VIM(1)

NAME
       vim - Vi IMproved, a programmers text editor
```

Whereas the ones that work, don't.

```
man(1)                                                                  man(1)

NAME
       man - format and display the on-line manual pages
```

```
yum(8)                                                                  yum(8)

NAME
       yum - Yellowdog Updater Modified
```

I've by no means checked all the manpages, but it does seem to be consistent. If I get time this weekend I'll see if I can figure out why this is.

**Comment by Konfekt** [comment]
Ok, if MANPAGER is nit defined then MAN_PN is unset. But why setting it explitly interfers, is yet mysterious to me.

**Comment by Konfekt** [comment]
Ah... Ok, pushed a commit so that now always the lowercase manpage is loaded if env MAN_PN =1. Can you try?

**Comment by kopischke** [comment]
Sorry to jump in, but I think I might be able to shed some light on the issues @verbitan is experiencing:
1. [The first issue](https://github.com/vim/vim/pull/491#issuecomment-217107645) seems to be caused by `s:MANPAGER` trying to access invalid regex match indices (it actually says so in the log posted, albeit obscured by the flurry of shell escape codes). [Line 27](https://github.com/Konfekt/vim/blob/18bdcecb9857cc5b484b32722b92eee5a3f6e430/runtime/plugin/manpager.vim#L27) should read
   
   ``` vim
   exe 'Man' page_sec[1] page_sec[0]
   ```
   
   to avoid this, as there is no index 3 – [line 14 only defines two capture groups](https://github.com/Konfekt/vim/blob/18bdcecb9857cc5b484b32722b92eee5a3f6e430/runtime/plugin/manpager.vim#L14) (the partial patterns defined above it use non-capturing groups `%()`).
2. [The second issue](https://github.com/vim/vim/pull/491#issuecomment-217418760) is probably due to either `man` or `:Man` expecting lower case man page names. Changing [line 22](https://github.com/Konfekt/vim/blob/18bdcecb9857cc5b484b32722b92eee5a3f6e430/runtime/plugin/manpager.vim#L22) to
   
   ``` vim
   let page_sec = matchlist(tolower(manpage), '^' . pagesec_pattern  . '$')
   ```
   
   should alleviate the issue.

**Comment by kopischke** [comment]
Ah, cross posting, I see @Konfekt is on it already. Do not mind me.

**Comment by verbitan** [comment]
@Konfekt your patch fixes the problem around loading manpages that start in uppercase :+1:

**Comment by Konfekt** [comment]
Ah, thanks Martin fir the clarification on capture groups. Just used %() interchangeably as better version of ()  as the doc shortly states.

**Comment by Konfekt** [comment]
Ok,  corrected according to @kopischke 's suggestions. Thanks !

**Comment by Konfekt** [comment]
@brammool Could you merge these corrections in?

**Comment by kopischke** [comment]
@Konfekt already merged in https://github.com/vim/vim/commit/91c4937be15b0b743b6bc495df602c1abbff6b87, unless I'm mistaken.

**Comment by Konfekt** [comment]
Ah, thanks. That was fast.

**Comment by Konfekt** [comment]
Dear @brammool , the index of the matchlist was off by one. Solved in https://github.com/Konfekt/vim/commit/159bfbe3ad3100cbe6a1533bb9cba7ae9c6ce07c

**Comment by kopischke** [comment]
@Konfekt ugh, my bad, sorry for that.

**Comment by Konfekt** [comment]
Mine too. Checked but did not realize the plugins in VIMRUNTIME are loaded before that in the home folder where the changes were made.

**Comment by brammool** [comment]
> Dear @brammool , the index of the matchlist was off by one. Solved in https://github.com/Konfekt/vim/commit/159bfbe3ad3100cbe6a1533bb9cba7ae9c6ce07c

Thanks, I'll include it.

## 

If you had to identify, in one word, the reason why the
human race has not achieved, and never will achieve, its
full potential, that word would be "meetings."

 /// Bram Moolenaar -- Bram@Moolenaar.net -- http://www.Moolenaar.net   \\
///        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ \\
\\  an exciting new programming language -- http://www.Zimbu.org        ///
 \\            help me help AIDS victims -- http://ICCF-Holland.org    ///
